docs/refs/api/rtos/callbacks.md

```diff
@@ -1,9 +1,184 @@
-#### Callbacks
+### Callbacks
 
-[Document about callbacks, their best practices and dos and don'ts.]
+A callback is a user provided function that a user may pass to an API. The callback allows the API to execute the user’s code in its own context.
 
-###### Thread example with callbacks
+```c++
+// Create a serial object
+Serial serial(USBTX, USBRX);
 
-The Callback API provides a convenient way to pass arguments to spawned threads.  
+// A function that echoes any received data back
+ void echo() {
+    while (serial.readable()) {
+         serial.putc(serial.getc());
+    }
+ }
+
+ int main(void) {
+     // Call our function echo whenever the serial line receives data
+     serial.attach(&echo, Serial::RxIrq);
+ }
+```
+
+The Callback class manages C/C++ function pointers so you don't have to. If you are asking yourself why you should use the Callback class, you should read the [Importance of State](callbacks_the_importance_of_state.md) documentation.
+
+### How to create callbacks
+
+First, you need to understand the syntax of the Callback type. The Callback type is a templated type parameterized by a C++ function declaration:
+
+``` c++
+// Callback</*return type*/(/*parameters*/)> cb;
+Callback<int(float)> cb;          // A callback that takes in a float and returns an int
+Callback<void(float)> cb;         // A callback that takes in a float and returns nothing
+Callback<int()> cb;               // A callback that takes in nothing and returns an int
+Callback<void(float, float)> cb;  // A callback that takes in two floats and returns nothing
+```
+
+You can create a Callback directly from a C function or function pointer with the same type:
+
+``` c++
+void dosomething(int) {
+    // do something
+}
+
+Callback<void(int)> cb(dosomething);
+```
+
+If an API provides a function that takes in a callback, you can pass in a C function or function pointer with the same type:
+
+``` c++
+class ADC {
+    void attach(Callback<void(float)> cb);
+};
+
+void dosomething(float) {
+    // do something
+}
+
+ADC adc;
+adc.attach(dosomething);
+```
+
+But what about state? The Callback type also supports passing a state pointer for a function. This state can be either a pointer to an object that is passed to a member function, or a pointer passed to a C-style function.
+
+Because this form of creating Callbacks requires two arguments, you need to create the Callback explicitly using the Callback constructor. The Callback also comes with the lowercase callback function, which creates callbacks based on the arguments type and avoids the need to repeat the template type.
+
+You can create a callback with a member function.
+
+``` c++
+class Thing {
+    int state;
+    void catinthehat(float f) {
+        state = // do something
+    }
+}
+
+// We can create a Callback with the Callback constructor
+Thing thing1;
+adc.attach(Callback<void(int)>(&thing1, &Thing::catinthehat));
```

Comment:
The param of catinthehat is float, so should it be:
`adc.attach(Callback<void(float)>(&thing1, &Thing::catinthehat));`

Reply:
I'll actually make catinthehat take int